control-plane/api-gateway/binding/validation.go

```diff
@@ -431,6 +431,53 @@ func routeAllowedForListenerHostname(hostname *gwv1beta1.Hostname, hostnames []g
 	return false
 }
 
+func keyFromLocalObjectReference(reference *gwv1beta1.LocalObjectReference) string {
+	return string(reference.Kind) + string(reference.Name)
+}
+
+// externalRefsOnRouteAllExist
+func externalRefsOnRouteAllExist(route *gwv1beta1.HTTPRoute, resources *common.ResourceMap) bool {
+	extensionRefs := resources.GetHTTPRouteExtensionRefs(route)
+	if len(extensionRefs) == 0 {
+		return true
+	}
+
+	for _, rule := range route.Spec.Rules {
+		for _, filter := range rule.Filters {
+			if filter.ExtensionRef == nil {
+				continue
+			}
+			filterKey := keyFromLocalObjectReference(filter.ExtensionRef)
+
+			_, ok := extensionRefs[filterKey]
```

Comment:
Rather than doing things like this, I think you probably just want something like a `resources.RouteExtensionRefExists(filter.ExtensionRef)` method.

As it is, this abstraction leaks, where you're creating a map in one context, assuming it's structured a particular way and then recreating the key here to access the elements of the map. Instead, it should be quicker (kind of trivial, but it'll save you at least one loop over the internal mapset) and less leaky to just leverage the internal resourcemap mapset semantics to check for the existence of the key within the context of the `resources.RouteExtensionRefExists` call.

Additionally, I think you should likely be checking the filter type first, and then checking for a corresponding `ExtensionRef`. If the filter type is set to use an extension, but then has no `ExtensionRef` field set, or an invalid `ExtensionRef.Kind` that we don't know how to handle, we should say, "nope, we're invalid".